### PR TITLE
HPCC-16030 Security Manager cache to implement read/write locks

### PR DESCRIPTION
**Pull request by RussWhitehead**
Currently critical sections are used to guard the LDAP cache. This PR
implements the usage of read/write locks, to allow concurrent reads
while no updates are taking place

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@rpastrana Please review

File: system/security/shared/caching.cpp [hunk 1]
```diff
@@ -173,29 +168,38 @@ void CResPermissionsCache::add( IArrayOf<ISecResource>& resources )
...
 
+//called from within a WriteLockBlock
 void CResPermissionsCache::removeStaleEntries(time_t tstamp)
 {
-    MapTimeStamp::iterator i; 
-    MapTimeStamp::iterator itL    = m_timestampMap.lower_bound(tstamp);
-    MapTimeStamp::iterator iBegin = m_timestampMap.begin();
-
-    for (i = iBegin; i != itL; i++)
+    int timeout = m_pParentCache->getCacheTimeout();
```
Comment (rpastrana) [inline system/security/shared/caching.cpp]: Doesn't seem like this should be fetched everytime, perhaps set this globally at load time?

File: system/security/shared/caching.cpp [hunk 2]
```diff
@@ -173,29 +168,38 @@ void CResPermissionsCache::add( IArrayOf<ISecResource>& resources )
...
+    if(timeout == 0 && m_pParentCache->isTransactionalEnabled())
+        timeout = 10; //Transactional timeout is set to 10 seconds for long transactions that might take over 10 seconds.
```
Comment (rpastrana) [inline system/security/shared/caching.cpp]: likewise this could be done at load time

File: system/security/shared/caching.cpp [hunk 3]
```diff
@@ -223,12 +227,27 @@ CPermissionsCache::~CPermissionsCache()
...
+int CPermissionsCache::lookup( ISecUser& sec_user, IArrayOf<ISecResource>& resources, bool* pFound)
 {
+    time_t tstamp;
+    time(&tstamp);
+
     const char* userId = sec_user.getName();
+
+    //First, clear stale cache entries for this CResPermissionsCache entry
+    {
+        WriteLockBlock writeLock(m_resPermCacheRWLock);
```
Comment (rpastrana) [inline system/security/shared/caching.cpp]: New logic minimizes serialization down to lines 239-246, is there a a way to minimize it even further?
We discussed only requesting writelocks for the case where the user entry has expired. Let us know why that approach isn't better than the current approach.

**Comment by rpastrana** [comment]
@RussWhitehead looks good. Left a couple of comments for your consideration.

**Comment by RussWhitehead** [comment]
@rpastrana Please review last commit, includes all your recommendations

**Comment by RussWhitehead** [comment]
@rpastrana Please rereview

**Comment by rpastrana** [comment]
@RussWhitehead looks good.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Comment by richardkchapman** [comment]
@ghalliday Please sanity-check

**Comment by ghalliday** [comment]
The code looks fine.  I suspect some of it could be simplified if the maps used Linked items for second - it might avoid the need to add specialist clean up code.

**Pull request merged**